utils/frame/remote-externalities/src/lib.rs

```diff
@@ -57,15 +57,60 @@ type ChildKeyValues = Vec<(ChildInfo, Vec<KeyValue>)>;
 
 const LOG_TARGET: &str = "remote-ext";
 const DEFAULT_HTTP_ENDPOINT: &str = "https://rpc.polkadot.io:443";
+const SNAPSHOT_VERSION: Compact<u16> = Compact(2);
+
 /// The snapshot that we store on disk.
 #[derive(Decode, Encode)]
 struct Snapshot<B: BlockT> {
+	snapshot_version: Compact<u16>,
 	state_version: StateVersion,
 	block_hash: B::Hash,
-	raw_storage: Vec<(H256, Vec<u8>)>,
+	// <Vec<Key, (Value, MemoryDbRefCount)>>
+	raw_storage: Vec<(H256, (Vec<u8>, i32))>,
 	storage_root: H256,
 }
 
+#[derive(Decode, Encode)]
+struct SnapshotOnlyVersion {
+	snapshot_version: Compact<u16>,
+}
+
+impl<B: BlockT> Snapshot<B> {
+	pub fn new(
+		state_version: StateVersion,
+		block_hash: B::Hash,
+		raw_storage: Vec<(H256, (Vec<u8>, i32))>,
+		storage_root: H256,
+	) -> Self {
+		Self {
+			snapshot_version: SNAPSHOT_VERSION,
+			state_version,
+			block_hash,
+			raw_storage,
+			storage_root,
+		}
+	}
+
+	fn load(path: &PathBuf) -> Result<Snapshot<B>, &'static str> {
+		let bytes = fs::read(path).map_err(|_| "fs::read failed.")?;
+		let maybe_version: Result<SnapshotOnlyVersion, _> = Decode::decode(&mut &*bytes);
```

Comment:
The `C` in SCALE stands for Concatenated, so if you put the `StorageVersion` at the end of the `Snapshot`, then you can just try to decode it after the `Snapshot` was decoded without any hacking. (probably need to change `bytes` to be `mut`)

Reply:
Ah I see what you mean, that would be much cleaner let me try.

Reply:
So I put `SnapshotVersion` at the *start* of `Snapshot`, and am able to decode and check it *before* trying to decode the rest of the bytes.

This isn't exactly what you described so I'm not sure if I'm missing something, let me know.

Reply:
Ah, once of the fields was changed, okay :facepalm:. Could you still put a magic number instead of relying on the fact that storage_version is < 2 to avoid a collision please?

Reply:
wdym put a magic number?

Reply:
Something unique that indicates that this is indeed a versioned snapshot and not a state_version https://en.wikipedia.org/wiki/Magic_number_(programming)  
But could be overkill.

Reply:
state_version could only ever have been 0 or 1 prior to this change (which is why we started the snapshot version at 2), so collision should be impossible